### ISSUE TITLE
TensorflowJs Incompatible shapes Error

### ISSUE THREAD

**Issue by Shanmax1996** [tensorflow/tensorflow#42408]
I build Unet architecture with TensorFlow Js.
But I faced a problem with Incompatible shapes error.
My input was [512, 512, 3] 3d tensor and label is [512, 512] 2d tensor.

my model was

    const input_layer = tf.input({ shape: [512, 512, 3], name: 'image_input'});
    
    const conv1 = tf.layers.conv2d({ filters: filter, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(input_layer);
    const batch1 = tf.layers.batchNormalization().apply(conv1);
    const act1 =  tf.layers.activation({activation: 'relu'}).apply(batch1);
    const conv11 = tf.layers.conv2d({ filters: filter, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(act1);
    const batch11 = tf.layers.batchNormalization().apply(conv11);
    const act11 =  tf.layers.activation({activation: 'relu'}).apply(batch11);
    const conv1_out = tf.layers.maxPooling2d({ poolSize: [2, 2] }).apply(act11);
    
    const conv2 = tf.layers.conv2d({ filters: filter*2, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(conv1_out);
    const batch2 = tf.layers.batchNormalization().apply(conv2);
    const act2 =  tf.layers.activation({activation: 'relu'}).apply(batch2);
    const conv22 = tf.layers.conv2d({ filters: filter*2, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(act2);
    const batch22 = tf.layers.batchNormalization().apply(conv22);
    const act22 =  tf.layers.activation({activation: 'relu'}).apply(batch22);
    const conv2_out = tf.layers.maxPooling2d({ poolSize: [2, 2] }).apply(act22);
    
    const conv3 = tf.layers.conv2d({ filters: filter*4, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(conv2_out);
    const batch3 = tf.layers.batchNormalization().apply(conv3);
    const act3 =  tf.layers.activation({activation: 'relu'}).apply(batch3);
    const conv33 = tf.layers.conv2d({ filters: filter*4, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(act3);
    const batch33 = tf.layers.batchNormalization().apply(conv33);
    const act33 =  tf.layers.activation({activation: 'relu'}).apply(batch33);
    const conv3_out = tf.layers.maxPooling2d({ poolSize: [2, 2] }).apply(act33);
    
    const conv4 = tf.layers.conv2d({ filters: filter*8, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(conv3_out);
    const batch4 = tf.layers.batchNormalization().apply(conv4);
    const act4 =  tf.layers.activation({activation: 'relu'}).apply(batch4);
    const conv44 = tf.layers.conv2d({ filters: filter*8, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(act4);
    const batch44 = tf.layers.batchNormalization().apply(conv44);
    const act44 =  tf.layers.activation({activation: 'relu'}).apply(batch44);
    const conv4_out = tf.layers.maxPooling2d({ poolSize: [2, 2] }).apply(act44);
    const conv4_dropout = tf.layers.dropout({rate: 0.5}).apply(conv4_out);
    
    const conv5 = tf.layers.conv2d({ filters: filter*16, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(conv4_dropout);
    const batch5 = tf.layers.batchNormalization().apply(conv5);
    const act5 =  tf.layers.activation({activation: 'relu'}).apply(batch5);
    const conv55 = tf.layers.conv2d({ filters: filter*16, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(act5);
    const batch55 = tf.layers.batchNormalization().apply(conv55);
    const act55 =  tf.layers.activation({activation: 'relu'}).apply(batch55);
    const conv5_dropout = tf.layers.dropout({rate: 0.5}).apply(act55);
    
    //Second part (up climb)
    const deconv6 = tf.layers.conv2dTranspose({ filters: filter*8, kernelSize: [3, 3], strides: [2, 2], padding: "same" }).apply(conv5_dropout);
    const deconv6_con = tf.layers.concatenate({axis: 3}).apply([deconv6, act44]);
    const deconv6_conv1 = tf.layers.conv2d({ filters: filter*8, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv6_con);
    const deconv6_batch1 = tf.layers.batchNormalization().apply(deconv6_conv1);
    const deconv6_act1 =  tf.layers.activation({activation: 'relu'}).apply(deconv6_batch1);
    const deconv6_conv11 = tf.layers.conv2d({ filters: filter*8, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv6_act1);
    const deconv6_batch11 = tf.layers.batchNormalization().apply(deconv6_conv11);
    const deconv6_act11 =  tf.layers.activation({activation: 'relu'}).apply(deconv6_batch11);
    const deconv6_dropout = tf.layers.dropout({rate: 0.5}).apply(deconv6_act11);
    
    const deconv7 = tf.layers.conv2dTranspose({ filters: filter*4, kernelSize: [3, 3], strides: [2, 2], padding: "same" }).apply(deconv6_dropout);
    const deconv7_con = tf.layers.concatenate({axis: 3}).apply([deconv7, act33]);
    const deconv7_conv1 = tf.layers.conv2d({ filters: filter*4, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv7_con);
    const deconv7_batch1 = tf.layers.batchNormalization().apply(deconv7_conv1);
    const deconv7_act1 =  tf.layers.activation({activation: 'relu'}).apply(deconv7_batch1);
    const deconv7_conv11 = tf.layers.conv2d({ filters: filter*4, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv7_act1);
    const deconv7_batch11 = tf.layers.batchNormalization().apply(deconv7_conv11);
    const deconv7_act11 =  tf.layers.activation({activation: 'relu'}).apply(deconv7_batch11);
    const deconv7_dropout = tf.layers.dropout({rate: 0.5}).apply(deconv7_act11);
    
    const deconv8 = tf.layers.conv2dTranspose({ filters: filter*2, kernelSize: [3, 3], strides: [2, 2], padding: "same" }).apply(deconv7_dropout);
    const deconv8_con = tf.layers.concatenate({axis: 3}).apply([deconv8, act22]);
    const deconv8_conv1 = tf.layers.conv2d({ filters: filter*2, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv8_con);
    const deconv8_batch1 = tf.layers.batchNormalization().apply(deconv8_conv1);
    const deconv8_act1 =  tf.layers.activation({activation: 'relu'}).apply(deconv8_batch1);
    const deconv8_conv11 = tf.layers.conv2d({ filters: filter*2, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv8_act1);
    const deconv8_batch11 = tf.layers.batchNormalization().apply(deconv8_conv11);
    const deconv8_act11 =  tf.layers.activation({activation: 'relu'}).apply(deconv8_batch11);
    
    const deconv9 = tf.layers.conv2dTranspose({ filters: filter, kernelSize: [3, 3], strides: [2, 2], padding: "same" }).apply(deconv8_act11);
    const deconv9_con = tf.layers.concatenate({axis: 3}).apply([deconv9, act11]);
    const deconv9_conv1 = tf.layers.conv2d({ filters: filter, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv9_con);
    const deconv9_batch1 = tf.layers.batchNormalization().apply(deconv9_conv1);
    const deconv9_act1 =  tf.layers.activation({activation: 'relu'}).apply(deconv9_batch1);
    const deconv9_conv11 = tf.layers.conv2d({ filters: filter, kernelSize: [3, 3], activation: "relu", padding: "same", kernelInitializer: "heNormal" }).apply(deconv9_act1);
    const deconv9_batch11 = tf.layers.batchNormalization().apply(deconv9_conv11);
    const deconv9_act11 =  tf.layers.activation({activation: 'relu'}).apply(deconv9_batch11);
    
    const output_layer = tf.layers.conv2d({ kernelSize: [1, 1], activation: "softmax", filters: 5}).apply(deconv9_act11);
    
    const model = tf.model({ inputs: input_layer, outputs: output_layer, name: "Unet" });
    
    model.compile({ loss: 'categoricalCrossentropy', optimizer: 'adam', metrics: ['accuracy'] });

and my train code was

    const xs = await tf.data.generator(data);        // input tensor3d [512, 512, 3]
    const ys = await tf.data.generator(labels);      // input tensor2d [512, 512]
    
    const ds = await tf.data.zip({xs, ys}).shuffle(2).batch(2);
    
    const model = await initModel();
    model.summary();
    
    await model.fitDataset(ds, {epochs: 5}).then(info => {
      console.log('Accuracy', info.history.acc);
    });

And I got this error message

    Incompatible shapes: [2,512,512] vs. [2,512,512,5]

**Comment by amahendrakar** [comment]
@Shanmax1996,
Issues related to TensorFlow.js are handled in the tensorflow/tfjs repo. Could you please submit a new issue from [this link](https://github.com/tensorflow/tfjs/issues/new) and fill in the template, so that we can track the issue there. Thanks!
